hbase-server/src/main/java/org/apache/hadoop/hbase/master/procedure/TruncateRegionProcedure.java

```diff
@@ -0,0 +1,260 @@
+    throws HBaseIOException {
+    super(env, hri);
+    this.region = hri;
+    checkOnline(env, region);
+  }
+
+  @Override
+  protected Flow executeFromState(final MasterProcedureEnv env, TruncateRegionState state)
+    throws InterruptedException {
+    if (LOG.isTraceEnabled()) {
+      LOG.trace(this + " execute state=" + state);
+    }
+    try {
+      switch (state) {
+        case TRUNCATE_REGION_PRE_OPERATION:
+          if (!prepareTruncate()) {
+            assert isFailed() : "the truncate should have an exception here";
+            return Flow.NO_MORE_STATE;
+          }
+          checkOnline(env, region);
+          assert region.getReplicaId() == RegionInfo.DEFAULT_REPLICA_ID || isFailed()
+            : "Can't truncate replicas directly. "
+              + "Replicas are auto-truncated when their primary is truncated.";
+          preTruncate(env);
+          setNextState(TruncateRegionState.TRUNCATE_REGION_MAKE_OFFLINE);
+          break;
+        case TRUNCATE_REGION_MAKE_OFFLINE:
+          addChildProcedure(createUnAssignProcedures(env));
+          setNextState(TruncateRegionState.TRUNCATE_REGION_REMOVE);
+          break;
+        case TRUNCATE_REGION_REMOVE:
+          deleteRegionsFromFileSystem(env);
+          setNextState(TruncateRegionState.TRUNCATE_REGION_MAKE_ONLINE);
+          break;
+        case TRUNCATE_REGION_MAKE_ONLINE:
+          addChildProcedure(createAssignProcedures(env));
+          setNextState(TruncateRegionState.TRUNCATE_REGION_POST_OPERATION);
+          break;
+        case TRUNCATE_REGION_POST_OPERATION:
+          postTruncate(env);
+          LOG.debug("truncate '" + getTableName() + "' completed");
+          return Flow.NO_MORE_STATE;
+        default:
+          throw new UnsupportedOperationException("unhandled state=" + state);
+      }
+    } catch (IOException e) {
+      if (isRollbackSupported(state)) {
+        setFailure("master-truncate-table", e);
+      } else {
+        LOG.warn("Retriable error trying to truncate table=" + getTableName() + " state=" + state,
+          e);
+      }
+    }
+    return Flow.HAS_MORE_STATE;
+  }
+
+  private TransitRegionStateProcedure createAssignProcedures(MasterProcedureEnv env)
+    throws IOException {
+    return createAssignProceduresForTruncateRegion(env, region);
+  }
+
+  private TransitRegionStateProcedure createUnAssignProcedures(MasterProcedureEnv env)
+    throws IOException {
+    return createUnassignProceduresForTruncateRegion(env, region);
+  }
+
+  private void deleteRegionsFromFileSystem(final MasterProcedureEnv env) throws IOException {
+    final MasterFileSystem mfs = env.getMasterServices().getMasterFileSystem();
+    final Path tableDir = CommonFSUtils.getTableDir(mfs.getRootDir(), getTableName());
+    final Configuration conf = env.getMasterConfiguration();
+    HRegionFileSystem.deleteRegionFromFileSystem(conf, mfs.getFileSystem(), tableDir, region);
```

Comment:
> This would delete the region dir entirely. I don't see where we recreate the region dir, later. Isn't this a problem?
While developing feature, I observed that subsequent task "TRUNCATE_REGION_MAKE_ONLINE" creates directory for region. So not added logic to create the directory explicitely.

Reply:
Thanks for pointing that. Yeah, that should be ok. Just one extra comment on the rollback: I see you are considering it only if we are in TRUNCATE_REGION_PRE_OPERATION. Couldn't we rollback from TRUNCATE_REGION_MAKE_OFFLINE too?

Reply:
> Thanks for pointing that. Yeah, that should be ok. Just one extra comment on the rollback: I see you are considering it only if we are in TRUNCATE_REGION_PRE_OPERATION. Couldn't we rollback from TRUNCATE_REGION_MAKE_OFFLINE too?
Thanks @wchevreuil. This is valid concern. I this its already handled [here](https://github.com/apache/hbase/blob/523e4f971d7c990c45574b203bdaf6d7935018d2/hbase-server/src/main/java/org/apache/hadoop/hbase/master/procedure/TruncateRegionProcedure.java#L139C1-L140C1)

Reply:
Thanks for pointing it out!